patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+                    "WHERE context <> 'internal' "
+                    "AND source IN ('configuration file', 'command line', 'environment variable') "
+                    "AND category <> 'Write-Ahead Log / Recovery Target' "
+                    "AND setting <> '(disabled)' "
+                    "OR name = ANY(%s)", (required_params,))
+
+        helper_dict = dict.fromkeys(['port', 'listen_addresses'])
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p == 'cluster_name' and v:
+                config['scope'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command',
+                       'hba_file', 'ident_file', 'config_file'):
+                # write commands to the local config due to security implications
+                # write hba/ident/config_file to local config to ensure they are not removed later
+                config['postgresql'].setdefault('parameters', {})
+                config['postgresql']['parameters'][p] = v
+            elif p in helper_dict:
+                helper_dict[p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
+
+    conn.close()
+
+    connect_port = parsed_dsn.get('port', os.getenv('PGPORT', helper_dict['port']))
+    config['postgresql']['connect_address'] = f'{get_ip()}:{connect_port}'
+    config['postgresql']['listen'] = f'{helper_dict["listen_addresses"]}:{helper_dict["port"]}'
+
+    # it makes sense to define postgresql.pg_hba/pg_ident only if hba_file/ident_file are set to defaults
+    default_hba_path = os.path.join(config['postgresql']['data_dir'], 'pg_hba.conf')
+    if config['postgresql']['parameters']['hba_file'] == default_hba_path:
+        try:
+            allowed_records = ['local', 'host', 'hostssl', 'hostnossl', 'hostgssenc', 'hostnogssenc']
+            if get_int_major_version(config) >= 16:
+                allowed_records += ['include', 'include_if_exists', 'include_dir']
+            with open(default_hba_path, 'r') as f:
+                config['postgresql']['pg_hba'] = list(filter(lambda i: i and i.split()[0] in allowed_records,
+                                                             (li.strip() for li in f.readlines())))
+        except OSError as e:
+            sys.exit(f'Failed to read pg_hba.conf: {e}')
+
+    default_ident_path = os.path.join(config['postgresql']['data_dir'], 'pg_ident.conf')
+    if config['postgresql']['parameters']['ident_file'] == default_ident_path:
+        try:
+            with open(default_ident_path, 'r') as f:
+                config['postgresql']['pg_ident'] = [i.strip() for i in f.readlines()
+                                                    if i.strip() and not i.startswith('#')]
```

Comment:
This code can bring unexpected lines because of `i.startswith('#')` with no stripping, e.g.:

```python
>>> with open('pg_ident.conf', 'r') as f:
...     print(f.read())
...
test1
 test2
# test3
 # test4
test5
>>> with open('pg_ident.conf', 'r') as f:
...     print([i.strip() for i in f.readlines() if i.strip() and not i.startswith('#')])
...
['test1', 'test2', '# test4', 'test5']
```